Improved economizer integration and single mode option for multi-speed coils

lgu1234 opened the pull request:
New feature: Economizer Integration Enhancements.

lgu1234 committed:
New feature check-in: Economizer Integration Enhancements

Added 2 new example files
Require transition

mjwitte committed:
Add new field for UnitarySystemPerformance:Multispee in ExpandObjects

lgu1234 committed:
Cleanup the code, remove unused variables, and add transition rules

lgu1234 committed:
Merge branch 'EconomizerIntegrationNew' of https://github.com/NREL/EnergyPlus into EconomizerIntegrationNew

Conflicts:
	src/ExpandObjects/epfilter.f90

lgu1234 committed:
Merge branch 'develop' into EconomizerIntegrationNew

Conflicts:
	tst/EnergyPlus/unit/MixedAir.unit.cc

lgu1234 committed:
Fix version differences between debug and release for a unit test file

Myoldmopar commented:
Hmm, looks like this replaces the other one actually, @mjwitte.

Also, why above does it say you pushed a commit 2 minutes ago?

![image](https://cloud.githubusercontent.com/assets/849824/12301151/d43034d6-b9db-11e5-9271-e8c1523e7e02.png)

That commit was on December 3.  By @lgu1234.  Methinks Github is having funny business.

mjwitte commented:
That's probably because I referenced this in my comment on the other commit.

mjwitte commented on design/FY2016/NFP-EcomonizerIntegration.md:
```diff
@@ -0,0 +1,1137 @@
+
+#### Field: Single Mode Operation
+
+This field specifies the coil operation mode for multiple speed DX cooling and heating coils. The allowed choice is Yes or No. The No choice allows a coil works between two adjacent speeds when a system load is greater than the coil capacity at speed 1. The Yes choice allows a coil works with a single capacity at a different speed. The speed number is determined by a system load.  
```
@lgu1234 Just to clarify.  This applies to each HVAC time step?  If that's correct, then I would suggest add to the first sentence " . . . and heating coils during each HVAC timestep."

lgu1234 replied:
Yes. You are correct. Good suggestion. The change is done in both NFP and 01c-InputOutputReference.md.

mjwitte commented on doc/src/docs/InputOutputReference/01c-InputOutputReference.md:
```diff
@@ -11374,6 +11379,10 @@ This field defines the number of heating speeds for the heat pump, and must matc
 
+#### Field: Single Mode Operation
+
+This field specifies the coil operation mode for multiple speed DX cooling and heating coils. The allowed choice is Yes or No. The No choice allows a coil works between two adjacent speeds when a system load is greater than the coil capacity at speed 1. The Yes choice allows a coil works with a single capacity at a different speed. The speed number is determined by a system load.  
```
" . . . and heating coils during each HVAC timestep."

lgu1234 replied:
Thanks. It is done in both documents.

mjwitte commented on doc/src/docs/InputOutputReference/01d-InputOutputReference.md:
```diff
@@ -15761,6 +15761,8 @@ In EnergyPlus the relief and outdoor air dampers, the economizer, and any outdoo
 
 Of course any type of setpoint manager can be used to establish a temperature setpoint at the mixed air node. But the Mixed Air Setpoint Manager is likely to be the most useful.
 
+When optional inputs of Cooling Coil Inlet Node Name, Cooling coil Outlet Node Name, and Minimum Temperature at Cooling Coil Outlet Node are provided, the setpoint temperature at the mixed air node is calculated based on the maximum of both the reference setpoint temperature and minimum cooling coil outlet temperature, minus the cooling coil air temperature reduction and the supply fan air temper ...
```
Typos in 2nd sentence "The setpoin temperature" and "claculate" and "cappedd" and "prevent cooling" --> "prevent the cooling"

lgu1234 replied:
Good catches. The typos are fixed.

mjwitte commented on doc/src/docs/InputOutputReference/01d-InputOutputReference.md:
```diff
@@ -15785,6 +15787,18 @@ The name of the supply fan outlet node.
+
+#### Field: Minimum Temperature at Cooling Coil Outlet Node
+
+In order to prevent cooling coil freezing, the minimum temperature at the cooling coil outlet is required based on requirement of ASHRAE Standard 90.1. The default value is 7.2C.
```
Suggest adding " . . . freezing during economizer operation,"

lgu1234 replied:
Done. Thanks.
@mjwitte I have a thought. When I search "economizer" over IO Ref, I can find words and field name containing economizer. However, I could not find any object name using Economizer. I assume two objects (Controller:OutdoorAir and OutdoorAir:Mixer) represent an economizer. Do you think we may need an object to represent an HVAC system component "Economizer"?

mjwitte replied:
@lgu1234 Given how large and complicated Controller:OutdoorAir has become, it might be useful to split off the economizer portions in to a separate object.  Something to consider if we reorganize the OA controls.

mjwitte commented on doc/src/docs/InputOutputReference/01d-InputOutputReference.md:
```diff
@@ -17236,6 +17247,8 @@ Note that the key value for these outputs is the AirLoopHVAC name, not the name
 
 * HVAC,Average,Air System Mixed Air Mass Flow Rate [kg/s]
 
+* HVAC,Average,Air System Ourdoor Air Maximum Flow Fraction []
```
Ourdoor --> Outdoor

lgu1234 replied:
Good catch. I think I corrected code and forgot to make correction in IO Ref. Thanks.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -52147,6 +52147,11 @@ UnitarySystemPerformance:Multispeed,
        \note Used only for Multi speed coils
        \note Enter the number of the following sets of data for air flow rates.
+  A2 , \field Single Mode Operation
+       \type choice
```
\note Controls coil operation during each HVAC timestep. 
\note Yes = operate at the highest speed possible without exceeding the current load.
\note No = allow operation at the average of two speeds to match the current load.
_Feel free to reword this_

lgu1234 replied:
Thanks for good additions. Here is my wording. Hope it makes sense.

```
   \note Controls coil operation during each HVAC timestep. 
   \note This choice does not apply to spped 1 opeation.
   \note Yes = operate at the higher speed possible without exceeding the current load.
   \note No = allow operation at the average of two adjacent speeds to match the current load.
```

mjwitte replied:
OK.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -71461,10 +71466,19 @@ SetpointManager:MixedAir,
        \type node
        \note Node(s) at which the temperature will be set
+  A7 , \field Cooling Coil Inlet Node Name
+       \type node
```
\note Optional field used to limit economizer operation to prevent freezing of DX cooling coil.
_Add same note to next two fields._

lgu1234 replied:
Done.

mjwitte commented on doc/src/docs/EngineeringReference/13b-EncyclopaedicRefs.md:
```diff
@@ -1811,6 +1811,8 @@ This model simulates the thermal performance of the indoor DX cooling coil, and
 
 When the model determines performance at Speed 1 (the lowest speed) or cycling between OFF and Speed 1, its performance is almost the same as the performance for the Coil:Cooling:DX:SingleSpeed model. However, the outlet conditions are calculated slightly differently. Therefore, the Coil:Cooling:DX:SingleSpeed model may be considered as a subset of the model described here. When the multispeed co ...
 
+When single mode operation is specified at higher speeds (above 1), defined in the UnitarySystemPerformance:Multispeed object, a child object of AirLoopHVAC:UnitarySystem, its performance is almost the same as the performance for the Coil:Cooling:DX:SingleSpeed model at different flow rate and capacity with given speed number. No liner interpolation is performed between two adjacent speeds.
```
Here and several similar places below.  Typo "liner" --> "linear".  Also, I would extend the last sentence: "No linear interpolation is performed between two adjacent speeds.  The highest speed which does not exceed the sensible load is used for each HVAC timestep."
_Or something like that._

lgu1234 replied:
Good catch for typos. Your suggestion is added. I would like to add one more sentence. "This operation does not apply to speed 1 operation and the case when the sensible load is above the highest sensible capacity."

mjwitte commented:
@lgu1234 @Myoldmopar I've reviewed docs and IDD.  Let me know if you need more.

lgu1234 committed:
Updated IO Ref., Eng. Ref, idd and NFP based on Mike's comments

lgu1234 commented:
@mjwitte I upload updated files based on your review. Could you also review the source code? Thanks.

nrel-bot-3 commented:
@Myoldmopar @lgu1234 @lgentile it has been 14 days since this pull request was last updated.

mjwitte committed:
Economizer integration - fix IDD typos

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into EconomizerIntegrationNew

Conflicts:
	src/EnergyPlus/HVACUnitarySystem.hh
	src/EnergyPlus/MixedAir.hh
	src/EnergyPlus/SetPointManager.cc
	src/EnergyPlus/SetPointManager.hh
	src/Transition/InputRulesFiles/Rules8-4-0-to-8-5-0.xls
	tst/EnergyPlus/unit/DXCoils.unit.cc
	tst/EnergyPlus/unit/HVACUnitarySystem.unit.cc

mjwitte commented:
@lgu1234 FYI, I've merged in develop and resolved conflicts.  Will start testing on the new example files.  You expect CI to have no diffs on existing test files, correct?

lgu1234 commented:
@mjwitte As I remembered, I expect to pass test of existing example files, except audit and eio diffs in a few files.

mjwitte committed:
Economizer integration - fix unit test

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into EconomizerIntegrationNew

mjwitte commented on src/EnergyPlus/MixedAir.cc:
```diff
@@ -4328,6 +4347,16 @@ namespace MixedAir {
 
+		if ( OAController( OAControllerNum ).CoolCoilFreezeCheck ) {
+			MaxOAFracBySetPoint = min( max( MaxOAFracBySetPoint, 0.0 ), 1.0 );
+			OAController( OAControllerNum ).MaxOAFracBySetPoint = MaxOAFracBySetPoint;
```
@lgu1234 When I run the new CoolingCoilFreezingPrevention tets file, the output always shows zero for "Air System Outdoor Air Maximum Flow Fraction".  
Also, as we move towards OO structs it's bad to have a local variable with the same name as a struct variable.  Suggest changing the name slightly or just delete the local variable and use OAController( OAControllerNum ).MaxOAFracBySetPoint instead.  @Myoldmopar Is that the proper guidance?

lgu1234 replied:
@mjwitte I ran the same input file using the source code before merge into develop and got non zero values on 4/1 and 4/2. Enclosed is a figure to show the same output variable.
![maxoafraction](https://cloud.githubusercontent.com/assets/5970656/13145654/1f436e7e-d620-11e5-9b6e-d7a0ced3a13e.jpg)

I am going to change a local variable name as l_MaxOAFracBySetPoint. Are you OK with the change?

mjwitte replied:
I don't see l_ as a prefix anywhere else in the code.  @Myoldmopar what's our preferred style for this?

mjwitte replied:
> I ran the same input file using the source code before merge into develop and got non zero values on 4/1 and 4/2. 

Oh.  So maybe my conflict resolution broke it.  I saved a pdf of the diffs before the merge.  I'll double-check against the current diffs.

mjwitte replied:
Merge is ok. Using your original test file with the maxOA limited to 0.2, there are values in Air System Outdoor Air Maximum Flow Fraction for some hours in April.  With my version of the test file without the 0.2 limit, these are all zero, but there are times where the OA fraction is different from the case without freezing protection. That's where I got confused.  And these happen at a time when the cooling coil isn't even running.
![image](https://cloud.githubusercontent.com/assets/4991075/13150919/615ee60a-d62d-11e5-83aa-70b6f2b4b106.png)

mjwitte commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -5312,10 +5320,19 @@ namespace SetPointManager {
 		FanInNode = this->FanInNode;
 		FanOutNode = this->FanOutNode;
 		RefNode = this->RefNode;
+		CoolCoilInNode = this->CoolCoilInNode;
```
@Myoldmopar Is this OK?  We've already got the same type of statement in the lines above.

Myoldmopar replied:
It's not great, but if it is functionally performing properly (no shadowing antics), then I don't think we should hold it up with today's freeze.  We could potentially note it as a clean-up to do post-freeze...

mjwitte commented:
@lgu1234 In the new CoolingCoilFreezingPrevention test file, the OA controller is limiting the max OA fraction to 0.2 which limits the effect of the new feature.  Suggest leaving this field blank in the Controller:OutdoorAir object
`MaxOAFractionSched;      !- Maximum Fraction of Outdoor Air Schedule Name`

When comparing this test file vs the same file with the freeze protection turned off (and the MaxOAFractionSched deleted) the runperiod in April is causing the cooling coil to operate due to the limited mixed air temperature while the non-freeze protection case is able to run on 100% OA and no DX coil.  My understanding of this limit is that it should only kick in if the cooling coil is operating.  Suggest that the OA controller logic check the compressor operation flag, or change the limit test to allow full economizer if it results in a mixed air temp that is equal to or below the mixed air setpoint.  Cycling fan mode complicates this a bit.

mjwitte commented:
@lgu1234 Looking at the other new test file,    .  If I take that file and change Single Mode Operation to No, it fails with this error:

```
   ** Severe  ** CalcMultiSpeedDXCoilCooling: Coil:Cooling:DX:MultiSpeed "HEAT PUMP ACDXCOIL 2 Developer error - inconsistent speed ratios.
   **   ~~~   ** CycRatio and SpeedRatio must be between 0.0 and 1.0
   **   ~~~   **  During Warmup, Environment=CHICAGO_IL_USA ANNUAL COOLING 1% DESIGN CONDITIONS DB/MCWB, at Simulation time=07/21 22:10 - 22:20
   **   ~~~   ** CycRatio=1.0, SpeedRatio = 1.1
   **  Fatal  ** Preceding condition(s) causes termination.
```

lgu1234 commented:
@mjwitte I know the failure. If you take a unitary system example file with unitary system template to make both heating and cooling coils as multispeed. It failed. I already talked to @rraustad before the new feature. Fortunately, when I used Single Mode Operation, I can bypass the failure. Therefore, this issue is not the economizer issue.

lgu1234 commented:
@mjwitte If I leave the field of Maximum Fraction of Outdoor Air Schedule Name blank without  MaxOAFractionSched, it will not have cooling, because 100% OA will be supplied to cool the space with very low OA temperature. In order to demonstrate freezing protection, I have to limit OA amount to trigger cooling, then reduce amount of OA to have freezing protection.

mjwitte commented:
But limiting the OA fraction to 0.2 misses the whole point of the feature.  There are some hours on the July day where the freeze protection kicks in in the early morning even without the 0.2 limit.

mjwitte commented:
Regarding the single mode file, if I change the heating to be single-speed, then it should run with single mode = No?

lgu1234 commented:
@mjwitte If single mode = No with a single speed heating coil, I assume it should work.

lgu1234 commented:
@mjwitte I tried to run a whole year and did not find any days to meet requirements. That is why I play with 0.2 with very low outdoor temperature.

mjwitte commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -5341,7 +5358,24 @@ namespace SetPointManager {
 
-		this->SetPt = Node( RefNode ).TempSetPoint - ( Node( FanOutNode ).Temp - Node( FanInNode ).Temp );
+		if ( CoolCoilInNode == 0 && CoolCoilOutNode == 0 ) {
+			this->SetPt = Node( RefNode ).TempSetPoint - ( Node( FanOutNode ).Temp - Node( FanInNode ).Temp );
```
@lgu1234 When FreezeCheckEnable is false, I would expect to get the standard mixed air setpoint like this line.  But the block below is always forcing MinTemp is the coil node names are present.

lgu1234 replied:
Mike:

I made code change slightly as follow:

```
    if ( CoolCoilInNode == 0 && CoolCoilOutNode == 0 ) {
        this->SetPt = Node( RefNode ).TempSetPoint - ( Node( FanOutNode ).Temp - Node( FanInNode ).Temp );
    } else {
        dtFan = Node( FanOutNode ).Temp - Node( FanInNode ).Temp;
        dtCoolCoil = Node( CoolCoilInNode ).Temp - Node( CoolCoilOutNode ).Temp;
        if ( dtCoolCoil > 0.0 && MinTemp > OutDryBulbTemp ) {
            this->FreezeCheckEnable = true;
        }
        if ( this->FreezeCheckEnable ) {
            if ( Node( RefNode ).Temp == Node( CoolCoilOutNode ).Temp ) { // blow through
                this->SetPt = max( Node( RefNode ).TempSetPoint, MinTemp ) - dtFan + dtCoolCoil;
            } else { // draw through
                if ( RefNode != CoolCoilOutNode ) { // Ref node is outlet node
                    this->SetPt = max( Node( RefNode ).TempSetPoint - dtFan, MinTemp ) + dtCoolCoil;
                } else {
                    this->SetPt = max( Node( RefNode ).TempSetPoint, MinTemp ) + dtCoolCoil;
                }
            }
        } else {
            this->SetPt = Node( RefNode ).TempSetPoint - ( Node( FanOutNode ).Temp - Node( FanInNode ).Temp );
        }
    }
```

When FreezeCheckEnable is false, the original calculation remains. However, differences of example runs are found.

lgu1234 replied:
@mjwitte Please see above changes. Are you OK with the changes.

mjwitte replied:
When you say "differences of example runs are found", what do you mean?  Differences in the CoolingCoilFreezePrevention file?  That would be OK.  Differences in other files would not.

Minor quibble with the code above, the "if ( this->FreezeCheckEnable)" isn't necessary, just let the prior if block continue.

lgu1234 replied:
How about this:

```
    this->SetPt = Node( RefNode ).TempSetPoint - ( Node( FanOutNode ).Temp - Node( FanInNode ).Temp );
    if ( CoolCoilInNode > 0 && CoolCoilOutNode > 0 ) {
        dtFan = Node( FanOutNode ).Temp - Node( FanInNode ).Temp;
        dtCoolCoil = Node( CoolCoilInNode ).Temp - Node( CoolCoilOutNode ).Temp;
        if ( dtCoolCoil > 0.0 && MinTemp > OutDryBulbTemp ) {
            this->FreezeCheckEnable = true;
            if ( Node( RefNode ).Temp == Node( CoolCoilOutNode ).Temp ) { // blow through
                this->SetPt = max( Node( RefNode ).TempSetPoint, MinTemp ) - dtFan + dtCoolCoil;
            } else { // draw through
                if ( RefNode != CoolCoilOutNode ) { // Ref node is outlet node
                    this->SetPt = max( Node( RefNode ).TempSetPoint - dtFan, MinTemp ) + dtCoolCoil;
                } else {
                    this->SetPt = max( Node( RefNode ).TempSetPoint, MinTemp ) + dtCoolCoil;
                }
            }
        }
    }
```

mjwitte replied:
That's better.

lgu1234 committed:
Made better setpoint calculation
Changed a local variable name to be different from a struct member name
All changes are based on Mike's comments

lgu1234 commented:
@mjwitte and @Myoldmopar Minor changes were uploaded. The differences are expected for CoolingCoilFreezingPrevention.idf.

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into EconomizerIntegrationNew

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into EconomizerIntegrationNew

Conflicts:
	tst/EnergyPlus/unit/MixedAir.unit.cc
